### DISCUSSION TITLE
Node Red: "Verbindung zum Server Verloren" nach Update

### DISCUSSION THREAD

**Discussion by Makiese** [OpenCCU/OpenCCU#1951] [🇩🇪 Fragen&Antworten]
Hallo in die Runde,

mit den letzten beiden Releases bekomme ich im Node-red (nutze RedMatic) dauerhaft folgende Meldung oben eingeblendet:

"Verbindung zum Server Verloren, Verbindung in x sec wieder hergestellt. Jetzt testen"

![verbindungsfehler](https://user-images.githubusercontent.com/42896364/191772747-878a5a57-7be9-4adc-b9fa-5e5f487ad9ba.PNG)

Aktuell verwendete Version mit Probleme:
![2022-09-22 16_23_32-RaspberryMatic WebUI](https://user-images.githubusercontent.com/42896364/191773281-349ae9a5-229e-44b8-bfa1-cc54e5c990cd.png)

Funktionierende Version ohne Probleme
![2022-09-22 16_44_05-RaspberryMatic WebUI](https://user-images.githubusercontent.com/42896364/191778354-733bd4da-a746-4f3e-8dcc-15b35d34cb4a.png)

Soweit lässt sich mit Node Red weiter arbeiten, jedoch habe ich Einschränkungen bei der Debug Ausgabe bemerkt. Diese funktioniert mit der neuen Version nicht mehr. 

Alles was hilft ist wieder auf die vorletzte Version zurückzuspringen. (Hatte zum Glück noch ein Snapshot)

Hat evtl. jemand ähnliche Probleme bemerkt?

Anmerkung:
Das löschen aller Flows in Node-Red, sowie die Aktualisierung von Node-Red auf v. 3.0.2 brachte keine Besserung.

**Comment by hammerandpick** [comment]
Die letzte Version, welche bei mir ohne diesen Fehler existierte, war die RaspberryMatic-3.63.8.20220330-rpi3. Seit der Version RaspberryMatic-3.63.9.20220625-rpi3 habe ich ebenfalls den Fehler.

**Comment by GCP007** [comment]
Hallo, 
zur Info & Fehlersuche, ich hab 2 Installationen laufen einmal mit & einmal ohne den beschriebenen Fehler:
- ohne Problem RPi4B mit NodeRed & auf der Konfig-Seite "ohne Projekt" in Betrieb
- mit Problem CCU3 mit NodeRed & auf der Konfig-Seite  "Projekte Aktiv" Theme "Standard"
Sonst gleiche Addons api/print/cuxd auf beiden Systemen und jeweils aktuelle Versionen

**Reply by hammerandpick** [reply]
In der Version RaspberryMatic-3.63.9.20220625-rpi3 hatte ich den Fehler, ohne dass Projekte aktiv war. Danach habe ich einen Rollback auf die vorhergehende Version gemacht. 
Auf dieser habe ich die Projekte aktiviert und später auf die 3.65.11.20221005 (rpi3) aktualisiert. Dort habe ich jetzt wieder den Fehler. Ich glaube daher nicht, dass es an den Projekten liegt.
Ich vermute, dass es irgend ein Problem mit dem Arbeitsspeicher / Speichermanagement gibt, welches zu dieser Fehlermeldung führt.  Der RPi4B wird mehr Speicher haben als die CCU3 und mein RPI3.

**Reply by hammerandpick** [reply]
Die relevanten Fehlermeldung im Log lauten bei mir vergleichbar dieser:
`Oct 24 13:39:50 homematic-raspi user.err monit[2007]: 'redmatic-nodered' mem amount of 314.0 MB matches resource limit [mem amount > 280 MB]`
bzw. dieser 
`Oct 24 13:30:57 homematic-raspi daemon.err node-red: (node:10567) UnhandledPromiseRejectionWarning: Error: socket hang up
Oct 24 13:30:57 homematic-raspi daemon.err node-red:     at connResetException (internal/errors.js:607:14)
Oct 24 13:30:57 homematic-raspi daemon.err node-red:     at Socket.socketOnEnd (_http_client.js:493:23)
Oct 24 13:30:57 homematic-raspi daemon.err node-red:     at Socket.emit (events.js:327:22)
Oct 24 13:30:57 homematic-raspi daemon.err node-red:     at endReadableNT (internal/streams/readable.js:1327:12)
Oct 24 13:30:57 homematic-raspi daemon.err node-red:     at processTicksAndRejections (internal/process/task_queues.js:80:21)
Oct 24 13:30:57 homematic-raspi daemon.err node-red: (node:10567) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
Oct 24 13:30:57 homematic-raspi daemon.err node-red: (node:10567) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
`

**Comment by Makiese** [comment]
Ich habe RaspberryMatic mit dem Addon Redmatic als VM auf einer Synology NAS laufen. Aber allem Anschein nach ist es unabhängig von der Plattform, wenn das Problem auch im Image 3.63.9.20220625-rpi3 existiert.

Ich hatte es auch schon mit einer neueren Node-Red Version probiert, weil die version von redmatic ist doch schon etwas alt und wird nicht mehr gepflegt. Aber es brachte keine Besserung. Auch alle Flows komplett löschen brachte keine Besserung.

**Comment by hammerandpick** [comment]
Bei mir laufen im Übrigen alle Flows weiterhin, auch das Dashboard zeigt plausible Werte an. Lediglich das Node-Red Frontend scheint nicht mehr ordnungsgemäß zu funktionieren.

**Comment by hammerandpick** [comment]
Ist übrigens als Issue #1892 von @jens-maus als nicht relevant für RaspberryMatic benannt worden.

**Reply by Makiese** [reply]
Hmm, dann wird das Problem wohl nie mehr gelöst werden. Denn RedMatic wird allem Anschein nach nicht mehr weiterentwickelt. 
Blöd nur, dass die Problematik erst mit Update von RaspberryMatic aufgekommen ist.

Ich werde bei Gelegenheit mal das aktuelle RaspberryMatic Image installieren und alle Flows in das neue Image überführen. Mal schauen ob das eine Lösung ist. Es ist nur immer etwas nervig alle Aktoren neu anzulernen.

**Comment by VoSs2o0o** [comment]
Bin auch betroffen: Update der CCU3 von original FW nach 3.65.11.20221005. Läuft im Firefox sauber, aber in Chrome/Vivaldi nicht. Chrome war bei mir gar nicht in Verwendung, Cache löschen, Private Seite funktioniert nicht. 
Im Entwicklertool regelmäßig "comms" mit wss://ccu3-webui/addons/red/comms Einträge unter Netzwerk im "defekten" Browser, in Firefox ohne diese Einträge.

**Reply by hammerandpick** [reply]
Ja genau das ist auch bei mir ein Fehler der Auftritt:
`Firefox kann keine Verbindung zu dem Server unter wss://SERVERNAME/addons/red/comms aufbauen. [red.min.js:16:25913](https://SERVERNAME/addons/red/red/red.min.js)` 

Nach einem erneuten Ab- und Anmelden habe ich noch folgende Fehler in der Konsole gefunden:
`XHRGET https://SERVERNAME/addons/red/settings?_=1682262716549 [HTTP/2 401 Unauthorized 9ms]`

`XML-Verarbeitungsfehler: Syntax-Fehler Adresse: https://SERVERNAME/addons/red/settings?_=1682262716549 Zeile Nr. 1, Spalte 1:settings:1:1`


im Edge habe ich dann folgendes gefunden:
`red.min.js:16  WebSocket connection to 'wss://SERVERNAME/addons/red/comms' failed: Error during WebSocket handshake: Unexpected response code: 400`
`s @ red.min.js:16`
`d @ red.min.js:16`
`success @ red.min.js:16`
`c @ vendor.js:2`
`fireWith @ vendor.js:2`
`l @ vendor.js:2`
`(anonym) @ vendor.js:2`

Vergleichbares gab es auch in diesem Forum: [Homematic-Forum](https://homematic-forum.de/forum/viewtopic.php?f=77&t=49957&sid=a1214f55962fded8c792db95ef8deffd&start=10) Damals war es ein Virenscanner, der die Übertragung der Logindaten verhindert hatte.

**Reply by hammerandpick** [reply]
Sporadisch hatte ich heute Verbindungen, aber nicht dauerhaft. 😞 Eine Regelmäßigkeit kann ich aber hierbei bislang nicht erkennen.

**Reply by VoSs2o0o** [reply]
..also ich verwende Kaspersky, und hatte schon öfters ein paar Problemchen mit Ihm, musste hier und da Ausnahmen schalten oder Sicherheitsfunktionen abschalten. Ich werde mal bei Gelegenheit in dieser Richtung forschen.

**Comment by grizzly0068** [comment]
Hallo zusammen,
ich bin heute auch gerade auf den Fehler aufgelaufen und war überrascht, denn ich kann meine Flows grad nicht bearbeiten.

Der Fehler erinnert mich an eine Konfiguration, die ich vor einiger Zeit für einen Reverse Proxy gebaut hatte, da war node-red und das dashboard hinter einem nginx reverse proxy. Das dashboard hatte funktioniert, die Konfiguration (node-red) nicht.
Ich musste dann für den webserver (nginx) eine socket forward directive in dessen Konfigurationsdatei eintragen. Genau weiß ich es nicht mehr, habe es damals nur abgeschrieben, aber vielleicht hilft es ja einem webserver admin beim suchen und fixen.
Denn wenn ich es richtig im Kopf habe, läuft für die GUI der RaspberryMatic auch ein webserver - glaube mini-httpd.

Hintergrund: das dashboard ist "stateless" im Webbrowser - http-request rein, http-response raus. Update des GUI ggfls. über Browser Push.
Die Konfiguration (node-red) braucht eine permanente Socket-Verbindung über TCP, d.h. der Proxy muss die TCP-Verbindung aufrecht erhalten. In einer Standard-Proxy-Konfiguration wird die TCP-Verbindung nach der Übertragung des response geschlossen. 
Bezogen auf auf node-red ist der GUI webserver auch ein Proxy, wenn er nicht gerade über das Filesystem sich die Seiten holt, denn node-red läuft standalone auf port 1880 mit einem eigenen Webserver-Interface, realisiert mit der node server class.

Wie gesagt, vielleicht nicht alles 100% korrekt beschrieben, ich bin kein Webadmin, nur User und Programmierer, aber vielleicht hilft die Idee irgendwem bei der Fehlersuche. Die oben beschriebene Konfiguration habe ich auch nicht weitergeführt, direkter Zugriff aus dem Heimnetz war schlussendlich leichter.
